mathics/builtin/manipulate.py

```diff
@@ -0,0 +1,276 @@
+        self._add_min_max_step_var(i, imin, imax, idefault, di, ilabel)
+
+    def apply_discrete_options_var1(self, i, options):
+        '{i_Symbol, options_List}'
+        if len(options.leaves) > 0:
+            self._add_discrete_options_var(i, options, options.leaves[0], i)
+
+    def apply_discrete_options_var2(self, i, options, idefault):
+        '{{i_Symbol, idefault_}, options_List}'
+        self._add_discrete_options_var(i, options, idefault, i)
+
+    def apply_discrete_options_var3(self, i, options, idefault, ilabel):
+        '{{i_Symbol, idefault_, ilabel_}, options_List}'
+        self._add_discrete_options_var(i, options, idefault, ilabel)
+
+
+class ListAnimate(Builtin):
+    rules = {
+        'ListAnimate[l_List]': 'Manipulate[Part[l, i], {i, 1, Length[l], 1}]'
+    }
+
+
+class Manipulate(Builtin):
+    """
+    <dl>
+    <dt>'Manipulate[$expr1$, $arg1$, ...]'
+        <dd>allows you to interactively compute and display an expression with different argument values.
+    </dl>
+
+    Manipulate[N[Sin[y]], {y, 1, 20, 2}]
+
+    Manipulate[i^3, {i, {2, x^4, a}}]
+
+    Manipulate[x^y, {x, 1, 20}, {y, 1, 3}]
+
+    """
+
+    attributes = ('HoldAll',)  # we'll call ReleaseHold at the time of evaluation below
+
+    messages = {
+        'noipywidget': 'Manipulate[] needs the ipywidgets module to work.',
+        'widgetargs': 'Illegal variable range or step parameters for ``.'
+    }
+
+    def apply(self, expr, args, evaluation):
+        'Manipulate[expr_, args__]'
+
+        if not _enabled:
+            evaluation.message('Manipulate', 'noipywidget')
+            return Symbol('Null')
+
+        try:
+            manip = Manipulations(evaluation)  # knows about the arguments and their widgets
+
+            for arg in args.get_sequence():
+                try:
+                    if not manip.dispatch(arg.evaluate(evaluation), evaluation):  # not a valid argument pattern?
+                        return Expression(self.get_name(), expr, *args.get_sequence())  # identity
+                except IllegalWidgetArguments as e:
+                    evaluation.message('Manipulate', 'widgetargs', _strip_namespace(str(e.var)))
+
+            clear_output_callback = evaluation.clear_output_callback
+            display_data_callback = evaluation.display_data_callback  # for pushing updates
+
+            def callback(**kwargs):
+                clear_output_callback(wait=True)
+
+                line_no = evaluation.definitions.get_line_no()
+
+                new_evaluation = Evaluation(evaluation.definitions, result_callback=display_data_callback,
+                                            out_callback=evaluation.out_callback)
+
+                vars = [Expression('Set', Symbol(name), value) for name, value in kwargs.items()]
+                evaluatable = Expression('ReleaseHold', Expression('Module', Expression('List', *vars), expr))
+                new_evaluation.evaluate([evaluatable], timeout=settings.TIMEOUT)
```

Comment:
This gives a timeout for each time you slide the slider, nice. I was worried that long-running computations would queue up (each with the default timeout) but it seems to interrupt them?